src/graph.py

```diff
@@ -635,6 +644,54 @@ def grid_search(self, prize_file, Gs, Bs, Ws):
         return forest, augmented_forest, params_by_nodes
 
 
+    def _eval_randomizations(self, params):
+        """
+        Convenience method which sets parameters and performs PCSF
+        """
+
+        self._reset_hyperparameters(params)
+        paramstring = "G_{}_B_{}_W_{}".format(*[int(x) if int(x) == x else x for x in [params['g'], params['b'], params['w']]])
+        logger.info("Randomizations for " + paramstring)
+
+        forest, augmented_forest = self.randomizations(params["noisy_edges_repetitions"], params["random_terminals_repetitions"])
+        
+        return paramstring, forest, augmented_forest
+
+
+    def _grid_randomization(self, prize_file, params):
+        """
+        Internal function which executes pcsf at every point in a parameter grid.
+        Subroutine of `grid_search`.
+
+        Arguments:
+            prize_file (str): filepath
+            params (dict): params with which to run the program
+
+        Returns:
+            results (dict)
+        """
+
+        # get number of cpus available to job
+        try:
+            n_cpus = int(os.environ["SLURM_JOB_CPUS_PER_NODE"])
+        except KeyError:
+            n_cpus = multiprocessing.cpu_count()
+
+        pool = multiprocessing.Pool(n_cpus)
+
+        self.prepare_prizes(prize_file)
+
+        model_params = [{'w': w, 'b': b, 'g':g} for (w, b, g) in product(params['w'], params['b'], params['g'])]
+        other_params = {key: params[key] for key in params if key not in 'wbg'} # `not in 'wbg'` is a shortened way of writing `key != 'w' & key != 'b' & key != 'g'`
+        all_params = [{**model_param, **other_params} for model_param in model_params]
+
+        results = pool.map(self._eval_randomizations, all_params)
+        # Convert to dictionary format
+        results = { paramstring: {"forest": forest, "augmented_forest": augmented_forest} for paramstring, forest, augmented_forest in results }
```

Comment:
Arguably this function should return the same format as the other functions in this section of graph.py. Maybe have another function (not a hidden one with a trailing `_`) call this function and then do the data transform? Not sure that would make sense here, just an option. What do you think?

Reply:
The other functions, namely `_eval_pcsf`, `_grid_pcsf`, `_grid_pcsf_parallel`, and `grid_search` are redundant (unless Tobi still uses `grid_search` for his visualization). 

I think it makes sense to output results as a dictionary keyed by `paramstring`. This would be especially useful if one wants to write a network produced by a specific set of parameters. If we removed the other 4 functions, we would would not have to worry about consistent output formats.

Reply:
075e1a45433dcde1858b5ab0fbe3db0539818962

Reply:
Is it possible to do a grid search without randomizations with the new function?